vcd/resource_vcd_catalog.go

```diff
@@ -37,6 +39,11 @@ func resourceVcdCatalog() *schema.Resource {
 				Optional: true,
 				ForceNew: true,
 			},
+			"storage_profile_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
```

Comment:
I tried to add this new parameter to my existing catalog. However, it failed with an API error.

Should it be possible to update `storage_profile_id` setting for an existing catalog without recreation?

Note: originally, I had created this catalog with Terraform, but then modified it with "share to all" setting through UI. Maybe that's why it's complaining now?

```
Terraform will perform the following actions:

  # vcd_catalog.shared_catalog will be updated in-place
  ~ resource "vcd_catalog" "shared_catalog" {
        created            = "2020-10-12T08:41:58.128Z"
        delete_force       = true
        delete_recursive   = true
        description        = "OS templates"
        id                 = "urn:vcloud:catalog:834cc462-e36f-47de-91db-95740f58026a"
        name               = "shared-catalog"
      + storage_profile_id = "urn:vcloud:vdcstorageProfile:020433e5-9186-41ff-9a34-df146b25c52c"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_catalog.shared_catalog: Modifying... [id=urn:vcloud:catalog:834cc462-e36f-47de-91db-95740f58026a]

Error: error updating catalog '': error updating catalog: API Error: 400: [ 4cd1f0ab-6c13-4bef-86f2-74c58ba3c39e ] HTTP 400 Bad Request
 - cvc-complex-type.2.4.a: Invalid content was found starting with element 'IsPublished'. One of '{"http://www.vmware.com/vcloud/v1.5":ExternalCatalogSubscriptionParams, "http://www.vmware.com/vcloud/v1.5":RemoteUriProbeResult, "http://www.vmware.com/vcloud/v1.5":PublishExternalCatalogParams}' is expected.
```

Reply:
p.s. I confirm that with a freshly created catalog, updating storage profile works as expected:

```
Terraform will perform the following actions:

  # vcd_catalog.sp_catalog will be updated in-place
  ~ resource "vcd_catalog" "sp_catalog" {
        created            = "2020-12-08T12:11:04.532Z"
        delete_force       = true
        delete_recursive   = true
        description        = "Catalog using custom storage profile"
        id                 = "urn:vcloud:catalog:778b6620-4f54-4d41-b295-a52b1ac5f490"
        name               = "sp-catalog"
      ~ storage_profile_id = "urn:vcloud:vdcstorageProfile:020433e5-9186-41ff-9a34-df146b25c52c" -> "urn:vcloud:vdcstorageProfile:d026cca5-5bbe-4e22-9ecc-a349948b9957"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_catalog.sp_catalog: Modifying... [id=urn:vcloud:catalog:778b6620-4f54-4d41-b295-a52b1ac5f490]
vcd_catalog.sp_catalog: Modifications complete after 5s [id=urn:vcloud:catalog:778b6620-4f54-4d41-b295-a52b1ac5f490]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

Reply:
Back on the topic of that API Error. I would like to try changing some other attribute in-place and see whether it throws the same error. So, I thought chaning `description`. To my surprise, it forces replacement. Maybe we could allow it to be updateable in-place as part of this PR?

```
Terraform will perform the following actions:

  # vcd_catalog.shared_catalog must be replaced
-/+ resource "vcd_catalog" "shared_catalog" {
      ~ created          = "2020-10-12T08:41:58.128Z" -> (known after apply)
        delete_force     = true
        delete_recursive = true
      ~ description      = "OS templates" -> "OS templates!" # forces replacement
      ~ id               = "urn:vcloud:catalog:834cc462-e36f-47de-91db-95740f58026a" -> (known after apply)
        name             = "shared-catalog"
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

Reply:
> Back on the topic of that API Error. I would like to try changing some other attribute in-place and see whether it throws the same error. So, I thought chaning description. To my surprise, it forces replacement. Maybe we could allow it to be updateable in-place as part of this PR?

I am looking at it as part of bug investigation now.

Reply:
>> Back on the topic of that API Error. I would like to try changing some other attribute in-place and see whether it throws the same error. So, I thought chaning description. To my surprise, it forces replacement. Maybe we could allow it to be updateable in-place as part of this PR?

> I am looking at it as part of bug investigation now.

`description` supports update now.

Reply:
> Should it be possible to update storage_profile_id setting for an existing catalog without recreation?

Yes. I just explicitly validated it by creating a catalog with 3.0.0 and then using 3.1.0 to set specific storage profile. This worked.

Next thing is to validate if issues occur because of having a shared catalog.

Reply:
> description supports update now.

Thanks, it works like a charm!!!

And I'm now able to confirm that the API Error is a generic update issue, not related to storage profile:

```
Terraform will perform the following actions:

  # vcd_catalog.shared_catalog will be updated in-place
  ~ resource "vcd_catalog" "shared_catalog" {
        created          = "2020-10-12T08:41:58.128Z"
        delete_force     = true
        delete_recursive = true
      ~ description      = "OS templates" -> "OS templates!"
        id               = "urn:vcloud:catalog:834cc462-e36f-47de-91db-95740f58026a"
        name             = "shared-catalog"
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_catalog.shared_catalog: Modifying... [id=urn:vcloud:catalog:834cc462-e36f-47de-91db-95740f58026a]

Error: error updating catalog '': error updating catalog: API Error: 400: [ c09ef084-d87a-413b-bf2b-982892513929 ] HTTP 400 Bad Request
 - cvc-complex-type.2.4.a: Invalid content was found starting with element 'IsPublished'. One of '{"http://www.vmware.com/vcloud/v1.5":ExternalCatalogSubscriptionParams, "http://www.vmware.com/vcloud/v1.5":RemoteUriProbeResult, "http://www.vmware.com/vcloud/v1.5":PublishExternalCatalogParams}' is expected.
```

Reply:
> And I'm now able to confirm that the API Error is a generic update issue, not related to storage profile:

After investigating UI I changed the `update`field to only send fields update deals with and skip the others. Looks like it helped and my manual testing allows updates to occur (but at the same time does not destroy manually set "Share" properties)

Reply:
Works like a charm now!